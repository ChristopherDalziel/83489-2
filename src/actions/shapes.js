import { v4 as uuidv4 } from "uuid";

export const addShape = ({
  sideOne = "",
  sideTwo = "",
  sideThree = "",
  sideFour = "",
} = {}) => ({
  type: "ADD_SHAPE",
  shape: {
    id: uuidv4(),
    sideOne,
    sideTwo,
    sideThree,
    sideFour,
  },
});
