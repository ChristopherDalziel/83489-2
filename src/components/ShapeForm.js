import React from "react";
import styled from "@emotion/styled";
import { connect } from "react-redux";
import {
  sortByTriangles,
  sortByCircles,
  sortByQuadrilaterals,
} from "../actions/filters";

const FormContainer = styled.div`
  padding: 20px 10px 0px 10px;
  color: black;
  @media (max-width: 815px) {
    width: 50vw;
  }
`;

const Instructions = styled.p`
  width: 50vw;
`;

export class ShapeForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      sideOne: props.shape ? props.shape.sideOne : "",
      sideTwo: props.shape ? props.shape.sideTwo : "",
      sideThree: props.shape ? props.shape.sideThree : "",
      sideFour: props.shape ? props.shape.sideFour : "",
      error: "",
    };
  }

  onSideOneChange = (e) => {
    const sideOne = e.target.value;
    if (!sideOne || sideOne.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideOne };
      });
    }
  };

  onSideTwoChange = (e) => {
    const sideTwo = e.target.value;
    if (!sideTwo || sideTwo.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideTwo };
      });
    }
  };

  onSideThreeChange = (e) => {
    const sideThree = e.target.value;
    if (!sideThree || sideThree.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideThree };
      });
    }
  };

  onSideFourChange = (e) => {
    const sideFour = e.target.value;

    if (!sideFour || sideFour.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideFour };
      });
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.sideOne) {
      this.setState(() => ({
        error: "Please provide at least 'Side One' ",
      }));
    } else if (
      this.state.sideOne &&
      this.state.sideTwo &&
      !this.state.sideThree
    ) {
      this.setState(() => ({
        error: "Please provide a shape with more or less than 2 sides",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        sideOne: this.state.sideOne,
        sideTwo: this.state.sideTwo,
        sideThree: this.state.sideThree,
        sideFour: this.state.sideFour,
      });

      if (
        this.state.sideOne &&
        this.state.sideTwo &&
        this.state.sideThree &&
        this.state.sideFour
      ) {
        this.props.sortByQuadrilaterals();
      } else if (
        this.state.sideOne &&
        this.state.sideTwo &&
        this.state.sideThree
      ) {
        this.props.sortByTriangles();
      } else if (this.state.sideOne) {
        this.props.sortByCircles();
      }
    }
  };

  render() {
    return (
      <FormContainer>
        <Instructions
          aria-readonly="true"
          role="Instructions on how to use this website"
        >
          Please enter your the lengths of each shape of the shape within the
          fields below, consider the lengths are been measured in pixels so you
          may want to use double digits to ensure your shape is visible!
          <Instructions>
            Then I will return your shape below, sorted into it's category!
          </Instructions>
        </Instructions>
        <form onSubmit={this.onSubmit}>
          {this.state.error && <p>{this.state.error}</p>}
          <label for="Shape Side One">Side One:</label>
          <input
            autoFocus
            onChange={this.onSideOneChange}
            value={this.state.sideOne}
            placeholder="Side One"
          ></input>
          <label for="Shape Side Two">Side Two:</label>
          <input
            disabled={!this.state.sideOne ? true : false}
            onChange={this.onSideTwoChange}
            value={this.state.sideTwo}
            placeholder="Side Two"
          ></input>
          <label for="Shape Side Three">Side Three:</label>
          <input
            disabled={!this.state.sideTwo ? true : false}
            onChange={this.onSideThreeChange}
            value={this.state.sideThree}
            placeholder="Side Three"
          ></input>
          <label for="Shape Side Four">Side Four:</label>
          <input
            disabled={!this.state.sideThree ? true : false}
            onChange={this.onSideFourChange}
            value={this.state.sideFour}
            placeholder="Side Four"
          ></input>
          <button>Create new shape</button>
        </form>
      </FormContainer>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    sortByTriangles: () => dispatch(sortByTriangles()),
    sortByCircles: () => dispatch(sortByCircles()),
    sortByQuadrilaterals: () => dispatch(sortByQuadrilaterals()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ShapeForm);
