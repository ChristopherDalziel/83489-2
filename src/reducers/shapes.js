const shapesReducerDefaultState = [];

const shapesReducer = (state = shapesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_SHAPE":
      return [...state, action.shape];
    case "RESET_SHAPES":
      return shapesReducerDefaultState;
    default:
      return state;
  }
};

export default shapesReducer;
