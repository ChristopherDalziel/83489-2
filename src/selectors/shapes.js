const expectedSides = {
  circles: 1,
  triangles: 3,
  quadrilaterals: 4,
};

const sortShapes = (shapes, { sortBy }) => {
  return shapes.filter((shape) => {
    const sides = Object.values(shape).filter((side) => side).length;
    return sides === expectedSides[sortBy];
  });
};

export default sortShapes;
