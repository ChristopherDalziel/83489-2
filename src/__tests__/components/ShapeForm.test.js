import React from "react";
import { shallow } from "enzyme";
import ShapeForm from "../../components/ShapeForm";
import shapes from "../fixtures/shapes";

import toJSON from "enzyme-to-json";

test("Should render shape form correctly", () => {
  const wrapper = shallow(<ShapeForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should render form with shape data", () => {
  const wrapper = shallow(<ShapeForm shape={shapes[0]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should render error for invalid form submission", () => {
  const wrapper = shallow(<ShapeForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
  wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  expect(wrapper.state("error").length).toBeGreaterThan(0);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("Should set sideOne on input change", () => {
  const value = "10";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(0).simulate("change", { target: { value } });
  expect(wrapper.state("sideOne")).toBe(value);
});

test("Should set sideTwo on input change", () => {
  const value = "15";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(1).simulate("change", { target: { value } });
  expect(wrapper.state("sideTwo")).toBe(value);
});

test("Should set sideThree on input change", () => {
  const value = "10";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(2).simulate("change", { target: { value } });
  expect(wrapper.state("sideThree")).toBe(value);
});

test("Should set sideFour on input change", () => {
  const value = "20";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(3).simulate("change", { target: { value } });
  expect(wrapper.state("sideFour")).toBe(value);
});

test("Should set sideOne if input is valid", () => {
  const value = "100.25";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(0).simulate("change", { target: { value } });
  expect(wrapper.state("sideOne")).toBe(value);
});

test("Should not set sideOne if input is invalid", () => {
  const value = "10.255";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(0).simulate("change", { target: { value } });
  expect(wrapper.state("sideOne")).toBe("");
});

test("Should set sideTwo if input is valid", () => {
  const value = "55.55";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(1).simulate("change", { target: { value } });
  expect(wrapper.state("sideTwo")).toBe(value);
});

test("Should not set sideTwo if input is invalid", () => {
  const value = "50.5510";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(1).simulate("change", { target: { value } });
  expect(wrapper.state("sideTwo")).toBe("");
});

test("Should set sideThree if input is valid", () => {
  const value = "40.45";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(2).simulate("change", { target: { value } });
  expect(wrapper.state("sideThree")).toBe(value);
});

test("Should not set sideThree if input is invalid", () => {
  const value = "29.015";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(2).simulate("change", { target: { value } });
  expect(wrapper.state("sideThree")).toBe("");
});

test("Should set sideFour if input is valid", () => {
  const value = "10";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(3).simulate("change", { target: { value } });
  expect(wrapper.state("sideFour")).toBe(value);
});

test("Should not set sideFour if input is invalid", () => {
  const value = "00.001";
  const wrapper = shallow(<ShapeForm />);
  wrapper.find("input").at(3).simulate("change", { target: { value } });
  expect(wrapper.state("sideFour")).toBe("");
});

test("Should call onSubmit prop for validation", () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ShapeForm shape={shapes[2]} onSubmit={onSubmitSpy} />
  );
  wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  expect(wrapper.state("error")).toBe("");
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    sideOne: shapes[2].sideOne,
    sideTwo: shapes[2].sideTwo,
    sideThree: shapes[2].sideThree,
    sideFour: shapes[2].sideFour,
  });
});
