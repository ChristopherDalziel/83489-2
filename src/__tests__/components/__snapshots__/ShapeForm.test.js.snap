// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render error for invalid form submission 1`] = `
<Styled(div)>
  <Styled(p)>
    Please enter your the lengths of each shape of the shape within the fields below, consider the lengths are been measured in pixels so you may want to use double digits to ensure your shape is visible!
    <Styled(p)>
      Then I will return your shape below, sorted into it's category!
    </Styled(p)>
  </Styled(p)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="sideOne"
    >
      Side One:
    </label>
    <input
      autoFocus={true}
      id="sideOne"
      onChange={[Function]}
      placeholder="Side One"
      value=""
    />
    <label
      for="sideTwo"
    >
      Side Two:
    </label>
    <input
      disabled={true}
      id="sideTwo"
      onChange={[Function]}
      placeholder="Side Two"
      value=""
    />
    <label
      for="sideThree"
    >
      Side Three:
    </label>
    <input
      disabled={true}
      id="sideThree"
      onChange={[Function]}
      placeholder="Side Three"
      value=""
    />
    <label
      for="sideFour"
    >
      Side Four:
    </label>
    <input
      disabled={true}
      id="sideFour"
      onChange={[Function]}
      placeholder="Side Four"
      value=""
    />
    <button
      role="button"
    >
      Create new shape
    </button>
  </form>
</Styled(div)>
`;

exports[`Should render error for invalid form submission 2`] = `
<Styled(div)>
  <Styled(p)>
    Please enter your the lengths of each shape of the shape within the fields below, consider the lengths are been measured in pixels so you may want to use double digits to ensure your shape is visible!
    <Styled(p)>
      Then I will return your shape below, sorted into it's category!
    </Styled(p)>
  </Styled(p)>
  <form
    onSubmit={[Function]}
  >
    <p>
      Please provide at least 'Side One' 
    </p>
    <label
      for="sideOne"
    >
      Side One:
    </label>
    <input
      autoFocus={true}
      id="sideOne"
      onChange={[Function]}
      placeholder="Side One"
      value=""
    />
    <label
      for="sideTwo"
    >
      Side Two:
    </label>
    <input
      disabled={true}
      id="sideTwo"
      onChange={[Function]}
      placeholder="Side Two"
      value=""
    />
    <label
      for="sideThree"
    >
      Side Three:
    </label>
    <input
      disabled={true}
      id="sideThree"
      onChange={[Function]}
      placeholder="Side Three"
      value=""
    />
    <label
      for="sideFour"
    >
      Side Four:
    </label>
    <input
      disabled={true}
      id="sideFour"
      onChange={[Function]}
      placeholder="Side Four"
      value=""
    />
    <button
      role="button"
    >
      Create new shape
    </button>
  </form>
</Styled(div)>
`;

exports[`Should render form with shape data 1`] = `
<Styled(div)>
  <Styled(p)>
    Please enter your the lengths of each shape of the shape within the fields below, consider the lengths are been measured in pixels so you may want to use double digits to ensure your shape is visible!
    <Styled(p)>
      Then I will return your shape below, sorted into it's category!
    </Styled(p)>
  </Styled(p)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="sideOne"
    >
      Side One:
    </label>
    <input
      autoFocus={true}
      id="sideOne"
      onChange={[Function]}
      placeholder="Side One"
      value={10}
    />
    <label
      for="sideTwo"
    >
      Side Two:
    </label>
    <input
      disabled={false}
      id="sideTwo"
      onChange={[Function]}
      placeholder="Side Two"
    />
    <label
      for="sideThree"
    >
      Side Three:
    </label>
    <input
      disabled={true}
      id="sideThree"
      onChange={[Function]}
      placeholder="Side Three"
    />
    <label
      for="sideFour"
    >
      Side Four:
    </label>
    <input
      disabled={true}
      id="sideFour"
      onChange={[Function]}
      placeholder="Side Four"
    />
    <button
      role="button"
    >
      Create new shape
    </button>
  </form>
</Styled(div)>
`;

exports[`Should render shape form correctly 1`] = `
<Styled(div)>
  <Styled(p)>
    Please enter your the lengths of each shape of the shape within the fields below, consider the lengths are been measured in pixels so you may want to use double digits to ensure your shape is visible!
    <Styled(p)>
      Then I will return your shape below, sorted into it's category!
    </Styled(p)>
  </Styled(p)>
  <form
    onSubmit={[Function]}
  >
    <label
      for="sideOne"
    >
      Side One:
    </label>
    <input
      autoFocus={true}
      id="sideOne"
      onChange={[Function]}
      placeholder="Side One"
      value=""
    />
    <label
      for="sideTwo"
    >
      Side Two:
    </label>
    <input
      disabled={true}
      id="sideTwo"
      onChange={[Function]}
      placeholder="Side Two"
      value=""
    />
    <label
      for="sideThree"
    >
      Side Three:
    </label>
    <input
      disabled={true}
      id="sideThree"
      onChange={[Function]}
      placeholder="Side Three"
      value=""
    />
    <label
      for="sideFour"
    >
      Side Four:
    </label>
    <input
      disabled={true}
      id="sideFour"
      onChange={[Function]}
      placeholder="Side Four"
      value=""
    />
    <button
      role="button"
    >
      Create new shape
    </button>
  </form>
</Styled(div)>
`;
